cdap-docs/developers-manual/source/building-blocks/workflows.rst

```diff
@@ -86,12 +94,206 @@ implement the ``run()`` method::
+      condition(new EnoughDataToProceed())
+        .condition(new IsWikipediaSourceOnline())
+          .addAction(new DownloadWikiDataAction())
+        .otherwise()
+          .addMapReduce(WikipediaPipelineApp.WIKIPEDIA_TO_DATASET_MR_NAME)
+        .end()
+        .addMapReduce(WikiContentValidatorAndNormalizer.NAME)
+        .fork()
+          .addSpark(SparkWikipediaAnalyzer.NAME)
+        .also()
+          .addMapReduce(TopNMapReduce.NAME)
+        .join()
+      .otherwise()
+      .end();
+    }
+
+.. _workflow-concurrent:
+
 Concurrent Workflows
 --------------------
 By default, a workflow runs sequentially. Multiple instances of a workflow can be run
 concurrently. To enable concurrent runs for a workflow, set its runtime argument
 ``concurrent.runs.enabled`` to ``true``.
 
+
+.. _workflow_token:
+
+Workflow Tokens
+===============
+
+In addition to passing the control flow from one node to the next, a **workflow token** is
+passed, available to each of the programs in the workflow. This allows programs to:
+
+- pass custom data (such as metrics, a status, or an error code) from one program in the 
+  workflow to subsequent programs;
+- query and set the data in the token;
+- fetch the data from the token which was set by a specific node; and
+- alter the job configuration based on a key in the token; for example, set a different
+  mapper/reducer class or a different input/output dataset for a Spark or MapReduce program.
+  
+The API is intended to allow appropriate action to be taken in response to the token, including
+logging and modifying the conditional execution of the workflow based on the token.
+
+Once a run is completed, you can query the tokens from past workflow runs for analyses that
+determine which node was executed more frequently and when. You can retrieve the token values
+that were added by a specific node in the workflow to debug the flow of execution.
+
+Scope
+-----
+Two scopes |---| *System* and *User* |---| are provided for workflow keys. CDAP adds keys
+(such as MapReduce counters) under the *System* scope. Keys added by user programs are
+stored under the *User* scope.
+
+Putting and Getting Token Values
+--------------------------------
+When a value is put into a token, it is stored under a specific key. Both keys and their
+corresponding values must be non-null. The token stores additional information about the 
+context in which the key is being set, such as the unique name of the workflow node. 
+
+To put a value into a token, first obtain access to the token from the workflow context, and
+then set a value for a specific key, as shown in this example from a MapReduce mapper::
+
+  public static class MyVerifier extends Mapper<LongWritable, Text, Text, NullWritable> {
+    public void map(LongWritable key, Text value, Context context)
+      throws IOException, InterruptedException {
+      String profile;
+      WorkflowToken workflowToken = context.getWorkflowToken();
```

Comment:
This is wrong. Mostly carried over from the wiki. The ``Context`` object here is a Hadoop class and has no idea about CDAP/workflow tokens. For accessing the workflow token in ``map()``/``reduce()`` methods, users will have to have their ``Mapper``/``Reducer`` classes implement ``ProgramLifecycle<MapReduceContext>``. After doing so, they can access workflow token in the ``initialize()``/``destroy`` methods. To access it in the ``map()``/``reduce()`` methods, they will have to cache a reference to the workflow token object as a class member in the ``initialize()`` method.

Example [here](https://github.com/caskdata/cdap/pull/3307/files#diff-aad4b0a2c7724edbf7d46f46590cd421R81)

The example doesn't quite cache it, because it doesn't need the token in the ``reduce`` method, but can be easily modified to store the workflow token as a class member. Infact, maybe we should add that to the example.